### ISSUE TITLE
Changes to coerce to ByteBuffer broke usage of InputStream

### ISSUE THREAD

**Issue by potetm** [cognitect-labs/aws-api#65]
It looks like you started requiring .markSupported InputStreams. This is understandable since you were fully consuming the InputStream into memory. However, my take is either:

* InputStreams should be supported without reading them fully into memory.
* InputStream should not supported at all. I know how to read into a byte array or a ByteBuffer, and not supporting streams at all would make the behavior clear.

**Comment by dchelimsky** [comment]
For which service/operation is this causing you trouble?

**Comment by potetm** [comment]
S3

**Comment by dchelimsky** [comment]
Confusion re: the api understood, and will be addressed.

That aside, the code was always reading the entire stream into a ByteBuffer before the change you're talking about, so I'm not clear on how this ever worked before the change in question. What kind of `InputStream` were you supplying?

**Comment by dchelimsky** [comment]
Actually, I guess it doesn't matter what kind it was. It would work if it fit in memory, and should have failed if it didn't.

**Comment by dchelimsky** [comment]
@potetm what is the concrete `InputStream` class that you're using that isn't `.markSupported`?

**Comment by dchelimsky** [comment]
We accept `InputStream` or byte array because we do intend to support streams larger than memory down the road and didn't want to change the API. Removing that now would be a breaking change for people already providing `InputStream`s that are `.markSupported`, and we don't want to do that.

`ByteBuffer` is an implementation detail, and we're not planning to surface that detail by accepting them.

I've updated the [README](/cognitect-labs/aws-api) and [Data Types](cognitect-labs/aws-api/blob/master/doc/types.md) doc to reflect that we only accept byte arrays and `.markSupported` `InputStream`s, and also improved the error message (coming in the next release).


I realize I asked you to open this issue (which I appreciate you doing), but I'm going to close it now and recommend you keep track of #14.

**Comment by potetm** [comment]
@dchelimsky Apologies. I didn't know you still wanted an answer to the question. I hit the issue when using `Files/newInputStream`, which uses a `sun.nio.ch.ChannelInputStream`. When I got the error message, I wrapped it in a `BufferedInputStream` and got a mark out of bounds error when the file size exceeded the buffer size.

Your decision makes sense. I would additionally recommend noting that, when using a buffered input stream, ensure that the buffer is large enough to hold the whole stream. Or perhaps just recommend `ByteArrayInputStream`?
